### PR TITLE
[EntryGeneration] Implement support for new godot array types

### PR DESCRIPTION
**Pull request by chippmann**
~~This does not fix multidimensional arrays yet.~~

### CODE REVIEW

**Review by piiertho — APPROVED** [review]
Thanks a lot for this awesome work.
Should wait for CI to be up before merging.

**Review by CedNaru — APPROVED** [review]

**Comment by chippmann** [comment]
@piiertho & @CedNaru please look at it again as i also added support for multidimensional arrays and enumVariantArrays.
@raniejade If you find time, I would sleep better if you had seen it as well as we did the defaultValue stuff together initially and its not really trivial especially with the enum stuff.

**Review by CedNaru — APPROVED** [review]

File: entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/generator/provider/ArrayRegistrationValuesHandler.kt
Comment (piiertho) [inline entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/generator/provider/ArrayRegistrationValuesHandler.kt]: Why not use `buildString { ... }` ?

**Review by piiertho — APPROVED** [review]
Seems good to me, maybe we can make code more readable with use of buildString, but that's a detail.

File: entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/extension/StringExt.kt
Comment (raniejade) [inline entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/extension/StringExt.kt]: I think we can collapse this into a single return:

```kotlin
"godot.core.VariantArray",
...
"godot.core.ObjectArray" -> true
```

File: entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/TypeToVariantAsClassNameMapper.kt
Comment (raniejade) [inline entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/TypeToVariantAsClassNameMapper.kt]:
```suggestion
                if (type?.isCompatibleList()) {
```

**Review by raniejade — CHANGES_REQUESTED** [review]

File: entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/extension/StringExt.kt
Comment (CedNaru) [reply entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/extension/StringExt.kt]: I think you mean GodotArray ? This is the true ancestor of all arrays, ObjectArray is just a specialized version for .... objects

File: entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/TypeToVariantAsClassNameMapper.kt
Comment (chippmann) [reply entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/TypeToVariantAsClassNameMapper.kt]: `type?.isCompatibleList()` returns a `Boolean?` but for an if statement you need a `Boolean`
So if `type?.isCompatibleList()` returns `null` it's `!= true`.
Or did i not get your point?

File: entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/extension/StringExt.kt
Comment (chippmann) [reply entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/extension/StringExt.kt]: > I think we can collapse this into a single return:
> 
> ```kotlin
> "godot.core.VariantArray",
> ...
> "godot.core.ObjectArray" -> true
> ```

Fixed it



> I think you mean GodotArray ? This is the true ancestor of all arrays, ObjectArray is just a specialized version for .... objects

When we get the `fqName` of that array, we don't get `GodotArray` we get `ObjectArray` that's why i check it here.

Comment (chippmann) [reply entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/extension/StringExt.kt]: hmm but maybe i could travers the parent list and see if any parent type is of type `GodotArray` let me check that.

File: entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/TypeToVariantAsClassNameMapper.kt
Comment (raniejade) [reply entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/TypeToVariantAsClassNameMapper.kt]: Ohhh, didn't realize that it returns `Boolean?`. Can `isCompatibleList` list return `Boolean` instead?

File: entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/extension/StringExt.kt
Comment (chippmann) [reply entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/extension/StringExt.kt]: > hmm but maybe i could travers the parent list and see if any parent type is of type `GodotArray` let me check that.

Yeah did it like that now

File: entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/TypeToVariantAsClassNameMapper.kt
Comment (chippmann) [reply entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/TypeToVariantAsClassNameMapper.kt]: It does but `type` can be null

Comment (chippmann) [reply entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/TypeToVariantAsClassNameMapper.kt]: other solution would be `type != null && type.isCompatibleList()`

Comment (raniejade) [reply entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/TypeToVariantAsClassNameMapper.kt]: @chippmann `type != null && type.isCompatibleList()` that is better and more readable too.

**Review by raniejade — APPROVED** [review]
LGTM, but one suggestion.

File: entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/TypeToVariantAsClassNameMapper.kt
Comment (chippmann) [reply entry-generation/godot-entry-generator/src/main/kotlin/godot/entrygenerator/mapper/TypeToVariantAsClassNameMapper.kt]: Changed it :-)

**Pull request merged**
